### PR TITLE
Implement LONGITUDINAL_DIFFUSION.

### PR DESCRIPTION
**Pull request by 1uc**

### CODE REVIEW

File: src/language/codegen.yaml [hunk 1]
```diff
@@ -87,7 +87,24 @@
...
                               - finalize_block:
                                   brief: "Statement block to be executed after calling linear solver"
-                                  type: StatementBlock                                  
+                                  type: StatementBlock
+                        - LongitudinalDiffusionBlock:
+                            brief: "Extracts information required for LONGITUDINAL_DIFFUSION for each KINETIC block."
+                            nmodl: "LONGITUDINAL_DIFFUSION_BLOCK"
+                            members:
+                              - name:
+                                  brief: "Name of the derivative block"
```
Comment (JCGoran) [inline src/language/codegen.yaml:96]: Should this be "derivative block", or something else?

File: src/visitors/longitudinal_diffusion_visitor.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    statements.reserve(nodes.size());
+    for (auto& node: nodes) {
+        statements.push_back(std::dynamic_pointer_cast<ast::Statement>(node));
+    }
+
+    return std::make_shared<ast::StatementBlock>(std::move(statements));
+}
+
+
+std::shared_ptr<ast::LongitudinalDiffusionBlock> create_block(ast::KineticBlock& node) {
```
Comment (JCGoran) [inline src/visitors/longitudinal_diffusion_visitor.cpp:34]: Should this be used externally? If not, it's missing a `static` specifier. If yes, it's missing a docstring.

File: src/codegen/codegen_info.cpp [hunk 3]
```diff
@@ -17,6 +19,48 @@ namespace codegen {
...
+}
+std::shared_ptr<ast::Expression> LongitudinalDiffusionInfo::diffusion_rate(
+    const std::string& index_name) const {
+    return substitute_index(index_name, rate_index_name, rate_expr);
+}
+
+double LongitudinalDiffusionInfo::dfcdc(const std::string& /* index_name */) const {
+    // Needed as part of the Jacobian to stabalize
+    // the implicit time-integration. However,
+    // currently, it's set to `0.0` for simplicity.
```
Comment (JCGoran) [inline src/codegen/codegen_info.cpp:44]: This comment can go in the header as it's not very obvious what the function does from the name alone.

File: src/codegen/codegen_neuron_cpp_visitor.hpp [hunk 4]
```diff
@@ -708,8 +723,8 @@ class CodegenNeuronCppVisitor: public CodegenCppVisitor {
 
     void visit_watch_statement(const ast::WatchStatement& node) override;
     void visit_for_netcon(const ast::ForNetcon& node) override;
-
-
+    void visit_longitudinal_diffusion_block(const ast::LongitudinalDiffusionBlock& node);
```
Comment (JCGoran) [inline src/codegen/codegen_neuron_cpp_visitor.hpp:726]: Mark `override` (if anything, to silence the GitHub warnings).

File: src/codegen/codegen_neuron_cpp_visitor.hpp [hunk 5]
```diff
@@ -708,8 +723,8 @@ class CodegenNeuronCppVisitor: public CodegenCppVisitor {
...
+    void visit_lon_diffuse(const ast::LonDiffuse& node);
```
Comment (JCGoran) [inline src/codegen/codegen_neuron_cpp_visitor.hpp:727]: Also mark `override`.

File: src/codegen/codegen_helper_visitor.cpp [hunk 6]
```diff
@@ -853,5 +855,62 @@ void CodegenHelperVisitor::visit_after_block(const ast::AfterBlock& node) {
     info.before_after_blocks.push_back(&node);
 }
 
+std::shared_ptr<ast::Compartment> find_compartment(const ast::LongitudinalDiffusionBlock& node,
```
Comment (JCGoran) [inline src/codegen/codegen_helper_visitor.cpp:858]: Is it used anywhere outside the translation unit? If not, mark `static`, if yes, add docstring.

File: src/codegen/codegen_info.hpp [hunk 7]
```diff
@@ -284,6 +284,31 @@ struct IndexSemantics {
         , size(size) {}
 };
 
+class LongitudinalDiffusionInfo {
```
Comment (JCGoran) [inline src/codegen/codegen_info.hpp:287]: (optional) Some brief docstrings for the public functions would be useful.

**Review by JCGoran — APPROVED** [review]
LGTM, mostly minor comments.

File: src/language/codegen.yaml (hunk 1 above)
Comment (1uc) [reply src/language/codegen.yaml:96]: ty

File: src/visitors/longitudinal_diffusion_visitor.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+namespace nmodl {
+namespace visitor {
+
+static std::shared_ptr<ast::StatementBlock> make_statement_block(
+    ast::KineticBlock& kinetic_block,
+    nmodl::ast::AstNodeType node_type) {
+    auto nodes = collect_nodes(kinetic_block, {node_type});
+
+    ast::StatementVector statements;
+    statements.reserve(nodes.size());
```
[the remark this answers is not in the document]
Comment (1uc) [reply src/visitors/longitudinal_diffusion_visitor.cpp:25]: It's the preferred pattern and appeases (my) static analysis.

File: src/visitors/longitudinal_diffusion_visitor.cpp (hunk 2 above)
Comment (1uc) [reply src/visitors/longitudinal_diffusion_visitor.cpp:34]: Correct, can be static.

File: src/codegen/codegen_info.cpp (hunk 3 above)
Comment (1uc) [reply src/codegen/codegen_info.cpp:44]: This one's tricky. I read a lot more code than doxygen documentation; as such I see the implementation more than the declaration. I would prefer to see the explanation right next to the `return 0.0` (which without comment is really strange).

One could completely inline `dfcdc`.

File: src/codegen/codegen_neuron_cpp_visitor.hpp (hunk 5 above)
Comment (1uc) [reply src/codegen/codegen_neuron_cpp_visitor.hpp:727]: ty

File: src/codegen/codegen_helper_visitor.cpp (hunk 6 above)
Comment (1uc) [reply src/codegen/codegen_helper_visitor.cpp:858]: ty

File: src/codegen/codegen_info.hpp (hunk 7 above)
Comment (1uc) [reply src/codegen/codegen_info.hpp:287]: done

File: src/codegen/codegen_info.cpp (hunk 3 above)
Comment (JCGoran) [reply src/codegen/codegen_info.cpp:44]: Fair enough, though in that case I think it's best to put this above the function signature so it shows up in the docs (I think doxygen then correctly picks it up). Also, possible small typo, should be "stabilize".

**Pull request merged**